### PR TITLE
Adding runtime packages if custom docker image is not provided

### PR DESCRIPTION
**Pull request by heemanshusuri**

### CODE REVIEW

File: R/environment.R [hunk 1]
```diff
@@ -206,6 +194,20 @@ generate_docker_file <- function(custom_docker_image = NULL,
   base_dockerfile <- NULL
   base_dockerfile <- paste0(base_dockerfile, sprintf("FROM %s\n",
                                                      custom_docker_image))
+  
+  base_dockerfile <- paste0(base_dockerfile, "RUN conda install -c r -y ",
+                            "r-essentials && conda clean -ay\n")
+  
+  base_dockerfile <- paste0(base_dockerfile, "RUN pip install --no-cache-dir ",
+                            "azureml-defaults\n")
```
Comment (ninghu) [inline R/environment.R]: merge to one docker image layer to save some space. also pin r-essentials to 3.6.0 for now

**Comment by ninghu** [comment]
This change is simpler that i thought 😄 please test our samples and make sure they are still working.

File: R/environment.R [hunk 2]
```diff
@@ -206,6 +194,20 @@ generate_docker_file <- function(custom_docker_image = NULL,
...
+  
+
```
Comment (ninghu) [inline R/environment.R]: remove the blank line

File: R/environment.R [hunk 3]
```diff
@@ -87,28 +87,16 @@ r_environment <- function(name, version = NULL,
   }
 
   if (is.null(custom_docker_image)) {
-    processor <- "cpu"
     if (use_gpu) {
-      processor <- "gpu"
+      base_image_with_address <- paste0("mcr.microsoft.com/azureml/base-",
```
Comment (ninghu) [inline R/environment.R]: Looks like `base_image_with_address `is not set when `custom_docker_image `is not null?

File: tests/testthat/test_environment.R [hunk 4]
```diff
@@ -32,32 +32,24 @@ test_that("create dockerfile", {
...
                                      cran_packages = c("ggplot2", "dplyr"))
-  expected_dockerfile <- paste0(
-    "FROM ubuntu-18.04\n",
-    "RUN R -e \"install.packages(\'ggplot2\', ",
-    "repos = \'http://cran.us.r-project.org\')\"\n",
-    "RUN R -e \"install.packages(\'dplyr\', ",
-    "repos = \'http://cran.us.r-project.org\')\"\n")
-  expect_equal(dockerfile, expected_dockerfile)
-  
+  expect_equal(dockerfile, "FROM ubuntu-18.04\n")
```
Comment (ninghu) [inline tests/testthat/test_environment.R]: We still can support incremental build with using a custom docker image as base image, right?

File: R/environment.R [hunk 5]
```diff
@@ -206,6 +194,20 @@ generate_docker_file <- function(custom_docker_image = NULL,
   base_dockerfile <- NULL
   base_dockerfile <- paste0(base_dockerfile, sprintf("FROM %s\n",
                                                      custom_docker_image))
+
```
Comment (ninghu) [inline R/environment.R]: if no extra packages are specified, can we just use user's custom docker image to start the job without image building?

File: tests/testthat/test_environment.R [hunk 6]
```diff
@@ -32,32 +32,24 @@ test_that("create dockerfile", {
...
+
   # github packages
-  dockerfile <- generate_docker_file(custom_docker_image = "ubuntu-18.04",
-                                     github_packages = c(
+  dockerfile <- generate_docker_file(github_packages = c(
                                        "https://github/user/repo1", 
                                        "https://github/user/repo2"))
   expected_dockerfile <- paste0(
-    "FROM ubuntu-18.04\n",
+    "FROM mcr.microsoft.com/azureml/base:openmpi3.1.2-ubuntu16.04\n",
```
Comment (ninghu) [inline tests/testthat/test_environment.R]: missing stpes for installing r sdk, azureml-defaults and other R packages.

File: R/environment.R (hunk 3 above)
Comment (heemanshusuri) [reply R/environment.R]: added

File: tests/testthat/test_environment.R (hunk 4 above)
Comment (heemanshusuri) [reply tests/testthat/test_environment.R]: yes right. made the changes

File: R/environment.R (hunk 5 above)
Comment (heemanshusuri) [reply R/environment.R]: yes, made relevant changes

File: R/environment.R [hunk 7]
```diff
@@ -199,14 +208,33 @@ container_registry <- function(address = NULL,
...
 #' @param custom_url_packages character vector of packages to be installed from
 #' local, directory or custom url.
+#' @param install_extra_packages logical parameter to specify if extra packages
+#' should be installed at runtime.
 generate_docker_file <- function(custom_docker_image = NULL,
                                  cran_packages = NULL,
                                  github_packages = NULL,
-                                 custom_url_packages = NULL) {
+                                 custom_url_packages = NULL,
+                                 install_extra_packages = TRUE) {
```
Comment (ninghu) [inline R/environment.R]: would it be better to call it `system_managed_packages`?

File: R/environment.R [hunk 8]
```diff
@@ -87,34 +87,43 @@ r_environment <- function(name, version = NULL,
...
+    else {
+      base_image_with_address <- paste0("mcr.microsoft.com/azureml/base:",
+                                        "openmpi3.1.2-ubuntu16.04")
     }
+
     env$docker$base_dockerfile <- generate_docker_file(base_image_with_address,
                                                        cran_packages,
                                                        github_packages,
                                                        custom_url_packages)
     env$docker$base_image <- NULL
```
Comment (ninghu) [inline R/environment.R]: If we moved this code outside if block, I am pretty sure we don't need else below.

File: R/environment.R (hunk 5 above)
Comment (ninghu) [reply R/environment.R]: I don't see this get addressed in latest changes. Did I miss something?

Comment (heemanshusuri) [reply R/environment.R]: it is. let's sync offline

File: R/environment.R (hunk 7 above)
Comment (heemanshusuri) [reply R/environment.R]: I renamed it to `install_system_packages`.

File: R/environment.R (hunk 8 above)
Comment (heemanshusuri) [reply R/environment.R]: as discussed offline, we need this block.

**Pull request merged**